File: Week_06/Practicum/Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def dfs(node):
    if node is None:
        return []

    return [node.val] + dfs(node.left) + dfs(node.right)

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        return dfs(root)

File: Week_06/Practicum/Binary Tree Preorder Traversal.cpp
class Solution {
public:
    void dfs(TreeNode* node, vector<int>& traversal) {
        if (!node) {
            return;
        }

        traversal.push_back(node->val);
        dfs(node->left, traversal);
        dfs(node->right, traversal);
    }

    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> traversal;
        dfs(root, traversal);
        return traversal;
    }
};
